### PR TITLE
Header network versions / Address, signature and keys versioning

### PR DESCRIPTION
**Pull request by Eitu33**

### CODE REVIEW

File: massa-versioning/src/lib.rs [hunk 1]
```diff
@@ -45,6 +45,8 @@
 //!
 //! Unit tests in versioning_factory.rs shows a basic but realistic implementation of a AddressFactory (impl the Factory trait)
 
+pub mod address_factory;
```
Comment (sydhds) [inline massa-versioning/src/lib.rs:48]: Maybe we can separate the factory implementations in a folder?

File: massa-signature/src/signature_impl.rs [hunk 2]
```diff
@@ -321,47 +490,76 @@ impl<'de> ::serde::Deserialize<'de> for KeyPair {
...
     fn partial_cmp(&self, other: &PublicKey) -> Option<Ordering> {
-        self.0.as_bytes().partial_cmp(other.0.as_bytes())
+        match (self, other) {
+            (PublicKey::PublicKeyV0(pubkey), PublicKey::PublicKeyV0(other_pubkey)) => {
+                pubkey.0.as_bytes().partial_cmp(other_pubkey.0.as_bytes())
+            }
+            (PublicKey::PublicKeyV0(pubkey), PublicKey::PublicKeyV1(other_pubkey)) => {
+                pubkey.0.as_bytes().partial_cmp(&other_pubkey.0.as_bytes())
+            }
+            (PublicKey::PublicKeyV1(pubkey), PublicKey::PublicKeyV0(other_pubkey)) => {
```
Comment (sydhds) [inline massa-signature/src/signature_impl.rs:528]: if used in BtreeMap (for instance), it is required to have total ordering which is not the case here...

File: massa-client/src/cmds.rs [hunk 3]
```diff
@@ -769,7 +769,8 @@ impl Command {
             Command::wallet_generate_secret_key => {
                 let wallet = wallet_opt.as_mut().unwrap();
 
-                let key = KeyPair::generate();
+                // NOTE: use factory? link with node?
```
Comment (sydhds) [inline massa-client/src/cmds.rs:772]: This is an important thing to discuss, so it's a friendly reminder :)

File: massa-consensus-worker/src/worker/init.rs [hunk 4]
```diff
@@ -41,7 +41,10 @@ pub fn create_genesis_block(
 ) -> Result<SecureShareBlock, ConsensusError> {
     let keypair = &cfg.genesis_key;
     let header = BlockHeader::new_verifiable(
+        // NEW TODO: what to implement here in case of restart?
```
Comment (sydhds) [inline massa-consensus-worker/src/worker/init.rs:44]: And more important question, what happen for a MipStore transitionning during a network shutdown?

File: massa-node/src/main.rs [hunk 5]
```diff
@@ -372,13 +372,27 @@ async fn launch(
         max_denunciations_per_block_header: MAX_DENUNCIATIONS_PER_BLOCK_HEADER,
     };
 
+    // Creates an empty default store
+    let mip_stats_config = MipStatsConfig {
+        block_count_considered: MIP_STORE_STATS_BLOCK_CONSIDERED,
+        counters_max: MIP_STORE_STATS_COUNTERS_MAX,
+    };
+    let mut mip_store =
+        MipStore::try_from(([], mip_stats_config)).expect("Cannot create an empty MIP store");
```
Comment (sydhds) [inline massa-node/src/main.rs:381]: It should not be empty anymore as we have versioned address no?

File: massa-signature/src/signature_impl.rs (hunk 2 above)
Comment (Eitu33) [reply massa-signature/src/signature_impl.rs:508]: `PublicKey` itself itself isn't, but `NodeId` is and it is nothing but a wrapper around a `PublicKey`:

```rust
/// `NodeId` wraps a public key to uniquely identify a node.
#[derive(
    Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd, SerializeDisplay, DeserializeFromStr,
)]
pub struct NodeId(PublicKey);
```

File: massa-versioning/src/lib.rs (hunk 1 above)
Comment (Eitu33) [reply massa-versioning/src/lib.rs:48]: I personally prefer good file naming over sub directories in a crate, if the team shares your opinion I don't mind changing it.

File: massa-versioning/src/versioning.rs [hunk 6]
```diff
@@ -88,9 +89,9 @@ machine!(
         /// Past start, can only go to LockedIn after the threshold is above a given value
         Started { pub(crate) threshold: Amount },
         /// Wait for some time before going to active (to let user the time to upgrade)
-        LockedIn { pub(crate) at: MassaTime },
+        LockedIn { pub(crate) delay: MassaTime },
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-versioning/src/versioning.rs]: Yes you're right I forgot to revert it, done now.

File: massa-versioning/src/versioning.rs [hunk 7]
```diff
@@ -20,6 +20,7 @@ use massa_time::MassaTime;
 #[repr(u32)]
 pub enum MipComponent {
     Address,
+    KeyPair,
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-versioning/src/versioning.rs:24]: Done

File: massa-versioning/src/versioning_factory.rs [hunk 8]
```diff
@@ -139,6 +137,24 @@ pub trait VersioningFactory {
             .collect()
     }
 
+    /// Get the version the current component with the given startegy
+    fn get_component_version_with_strategy(
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-versioning/src/versioning_factory.rs:141]: Done

File: massa-node/src/main.rs (hunk 5 above)
Comment (Eitu33) [reply massa-node/src/main.rs:381]: Was planning to update this when I move to high end testing, didn't get there yet.

File: massa-consensus-worker/src/worker/init.rs (hunk 4 above)
Comment (sydhds) [reply massa-consensus-worker/src/worker/init.rs:44]: Added to #3631

File: massa-client/src/cmds.rs (hunk 3 above)
Comment (sydhds) [reply massa-client/src/cmds.rs:772]: Added to #3631

File: massa-signature/src/signature_impl.rs (hunk 2 above)
Comment (Eitu33) [reply massa-signature/src/signature_impl.rs:508]: For info: what was asked here was to have ordering of versions too

Comment (Eitu33) [reply massa-signature/src/signature_impl.rs]: There is actually no need to explicitly specify the version ordering, it is done naturally by the initial version byte, you can see the update in https://github.com/massalabs/massa/pull/3876/commits/455691a6135627e7c487a33ce3d666492b552fbd

**Pull request merged**